src/GeolocationContext.tsx

```diff
@@ -143,39 +152,46 @@ export const GeolocationContextProvider: React.FC = ({children}) => {
     }
   }
 
+  const watchIdRef = useRef<number>(-1);
+
   const hasPermission = state.status === 'granted' && state.locationEnabled;
-  useEffect(() => {
-    let watchId: number;
 
-    async function startLocationWatcher() {
-      if (!hasPermission) {
-        dispatch({
-          type: 'LOCATION_CHANGED',
-          position: null,
-          locationName: geoLocationName,
-        });
-      } else {
-        const config: GeoOptions = {
-          enableHighAccuracy: true,
-          distanceFilter: 20,
-        };
+  const startLocationWatcher = useCallback(async () => {
+    if (!hasPermission) {
+      dispatch({
+        type: 'LOCATION_CHANGED',
+        position: null,
+        locationName: geoLocationName,
+      });
+    } else {
+      const config: GeoOptions = {
+        enableHighAccuracy: true,
+        distanceFilter: 20,
+      };
 
-        watchId = Geolocation.watchPosition(
-          (position) => {
-            dispatch({
-              type: 'LOCATION_CHANGED',
-              position,
-              locationName: geoLocationName,
-            });
-          },
-          handleLocationError,
-          config,
-        );
-      }
+      watchIdRef.current = Geolocation.watchPosition(
+        (position) => {
+          dispatch({
+            type: 'LOCATION_CHANGED',
+            position,
+            locationName: geoLocationName,
+          });
+        },
+        handleLocationError,
+        config,
+      );
     }
+  }, [hasPermission, geoLocationName, dispatch]);
+
+  const stopLocationWatcher = () => Geolocation.clearWatch(watchIdRef.current);
+
+  useEffect(() => {
     startLocationWatcher();
-    return () => Geolocation.clearWatch(watchId);
-  }, [hasPermission]);
+    return stopLocationWatcher;
+  }, []);
+
+  useOnResume(startLocationWatcher);
+  useOnPause(stopLocationWatcher);
```

Comment:
In general there is quite a bit more risk when doing it like this than just doing it all in one effect, like this (pseudocode):
```tsx
const isFocusedAndActive = useIsFocuesAndActive();
useEffect(() => {
  if (isFocusedAndActive) {
    const id = setupWatcher():
    return () => clearWatch(id)
  }
}, [isFocusedAndActive])
```

We can take a quick discussion regarding risk/benefit.

Reply:
Currently refactored like this:

```ts
  useEffect(() => {
    if (appStatus === 'active') {
      if (!hasPermission) {
        updateLocation(null, geoLocationName);
      } else {
        const watchId = Geolocation.watchPosition(
          (position) => updateLocation(position, geoLocationName),
          handleLocationError,
          config,
        );
        return () => Geolocation.clearWatch(watchId);
      }
    }
  }, [geoLocationName, hasPermission, appStatus]);
```